### PR TITLE
Add some docblocks for the Storage layer

### PR DESCRIPTION
**Pull request by puiterwijk**

### CODE REVIEW

File: server/utils/general.go [hunk 1]
```diff
@@ -2,20 +2,28 @@ package utils
...
+// ReadClosers' when done reading a single Reader.
+// NOTE: This should only be used if the outer Close() function does *not* call the inner Close() function,
+// like is the case with zlib.Reader and gzip.Reader.
 type innerReadCloser struct {
 	closeOuter bool
 	inner      io.ReadCloser
 	outer      io.ReadCloser
 }
 
+// NewInnerReadCloser returns a new innerReadCloser structure, with the indicated inner and outer readers
```
Comment (bowlofeggs) [inline server/utils/general.go:15]:
```suggestion
// NewInnerReadCloser returns a new innerReadCloser structure, with the indicated inner and outer readers
// If set to true, closeOuter will close the outer reader when NewInnerReadCloser.Close() is used.
```

File: server/storage/generic.go [hunk 2]
```diff
@@ -9,38 +9,57 @@ import (
 	"io"
 )
 
+// ObjectID is a Git Object ID (sha1 digest in hex format without prefix)
 type ObjectID string
 
+// ZeroID is the "null" object ID, indicating there's no object here
```
Comment (bowlofeggs) [inline server/storage/generic.go:15]: I assume this is used in ```git``` this way? If so, maybe we could mention that here too.

File: server/storage/generic.go [hunk 3]
```diff
@@ -9,38 +9,57 @@ import (
...
 
+// ObjectIDLength defines the length of an Object ID
 const ObjectIDLength = 40
 
+// ObjectIDFromRaw parses the bytes of a sha1 digest and returns the standard object ID format representation
 func ObjectIDFromRaw(rawid []byte) ObjectID {
 	return ObjectID(hex.EncodeToString(rawid))
 }
 
-type Reference string
```
Comment (bowlofeggs) [inline server/storage/generic.go:23]: Is this unused? Perhaps break this change into its own commit, if so.

File: server/storage/generic.go [hunk 4]
```diff
@@ -9,38 +9,57 @@ import (
...
-
+// ObjectType represents the type of a Git Object
+// In Pack files, this is represented by 3 bits.
 type ObjectType int
 
 const (
+	// ObjectTypeBad is an invalid object type and should result in an error
 	ObjectTypeBad    ObjectType = -1
```
Comment (bowlofeggs) [inline server/storage/generic.go]: Is this one not used by ```git```, since it is outside of the range of 3 bits as well (like 9 below)?

File: server/storage/generic.go [hunk 5]
```diff
@@ -9,38 +9,57 @@ import (
...
+	// ObjectTypeOfsDelta encodes an "offset delta" object
+	// These are deltas that refer to another object in the same pack by offset from the base
 	ObjectTypeOfsDelta ObjectType = 6
+	// ObjectTypeRefDelta encodes a "reference delta" object
+	// These are deltas that refer to another object in the same repository by encoding the base object ID
 	ObjectTypeRefDelta ObjectType = 7
+	// ObjectTypeAny should never occur, as it's an encoded value for "any" object (anything goes), which we don't want
 	ObjectTypeAny      ObjectType = 8
+	// ObjectTypeMax should never occur, as it's higher than available in the 3 bits reserved for object types
 	ObjectTypeMax      ObjectType = 9
```
Comment (bowlofeggs) [inline server/storage/generic.go]: Why define this if it will never occur? Obviously, it is defined outside of the scope of this PR, so this isn't really a comment on the PR itself, but a general question.

File: server/storage/generic.go [hunk 6]
```diff
@@ -76,36 +95,62 @@ func ObjectTypeFromHdrName(hdrname string) ObjectType {
...
+	// GetProjectStorage returns a scoped instance for a particular project
 	GetProjectStorage(project string) ProjectStorageDriver
 }
 
+// ErrObjectNotFound is an error encoding an object was not found by the storage driver
 var ErrObjectNotFound = errors.New("Object not found")
 
+// ProjectStorageDriver is a scoped storage driver instance to a single project
+// SECURITY PROPERTY: a ProjectStorageDriver should not be able to find any objects that were never pushed to this specific project,
+// so if object A was pushed to project X but not Y, running a ReadObject() from a PRojectStorageDriver scoped to project Y *must*
```
Comment (bowlofeggs) [inline server/storage/generic.go]:
```suggestion
// so if object A was pushed to project X but not Y, running a ReadObject() from a ProjectStorageDriver scoped to project Y *must*
```

File: server/storage/generic.go [hunk 7]
```diff
@@ -76,36 +95,62 @@ func ObjectTypeFromHdrName(hdrname string) ObjectType {
...
+// return ErrObjectNotFound. This avoids possible attacks for getting objects from repositories an attacker should not have access to:
+// - Perform a Pull operation where the "want" list contains the object ID
+// - Perform a Push where one of the Commit objects points to the other object as its Parent or Tree, which would result in us returning
+//   that object on the next pull
 type ProjectStorageDriver interface {
+	// ReadObject is used to read the Object objectid, and returns the objects' type, size and a ReadCloser (make sure to close after reading)
```
Comment (bowlofeggs) [inline server/storage/generic.go]:
```suggestion
	// ReadObject is used to read the Object objectid, and returns the object's type, size and a ReadCloser (make sure to close after reading)
```

File: server/storage/generic.go [hunk 8]
```diff
@@ -76,36 +95,62 @@ func ObjectTypeFromHdrName(hdrname string) ObjectType {
...
 }
 
+// StagedObject is an object that is currently being written to storage
+// NOTE: Finalize *must* be called before Close is called! Otherwise the underlying storage implementation must toss the object
+// and regard it as "incomplete".
 type StagedObject interface {
 	io.Writer
 	io.Closer
+	// Finalize informs the storage driver the full object was written via the Write function, and an object ID can be computed.
+	// The provided objid will be compared to the computed object ID, and if they don't match, the Finalize returns an error,
```
Comment (bowlofeggs) [inline server/storage/generic.go]:
```suggestion
	// The provided objid will be compared to the computed object ID, and if they don't match, Finalize returns an error.
```

File: server/storage/generic.go [hunk 9]
```diff
@@ -76,36 +95,62 @@ func ObjectTypeFromHdrName(hdrname string) ObjectType {
...
+	// this can be used if the object ID is known, so we can be sure that the correct object was written to disk.
```
Comment (bowlofeggs) [inline server/storage/generic.go]:
```suggestion
	// This can be used if the object ID is known, so we can be sure that the correct object was written to disk.
```

File: server/storage/generic.go [hunk 10]
```diff
@@ -147,10 +193,13 @@ func (w *oidWriter) Close() error {
 	return w.inner.Close()
 }
 
+// getObjectID computes the Object ID of the object that was written via the oidWriter
```
Comment (bowlofeggs) [inline server/storage/generic.go]:
```suggestion
// getObjectID computes the Object ID of the object that was written via the oidWriter.
```

File: server/storage/generic.go [hunk 11]
```diff
@@ -147,10 +193,13 @@ func (w *oidWriter) Close() error {
...
+// this must be called after Close() is called, to make sure buffers are flushed.
```
Comment (bowlofeggs) [inline server/storage/generic.go]:
```suggestion
// This must be called after Close() is called, to make sure buffers are flushed.
```

File: server/storage/tree.go [hunk 12]
```diff
@@ -21,6 +21,13 @@ const (
 	compressMethodZlib                = 2
 )
 
+// treeStorageDriverInstance is the StorageDriver for "tree" storage.
+// This storage driver stores every object of every project in a directory <basedir>/<projectname>/<objid[0:2]>/<objid[2:]><extension>
+// where <baseid> is configured, <projectname> is the project name used for the object, <objid[0:2]> are the first two characters
```
Comment (bowlofeggs) [inline server/storage/tree.go]:
```suggestion
// where <basedir> is configured, <projectname> is the project name used for the object, <objid[0:2]> are the first two characters
```

File: server/storage/tree.go [hunk 13]
```diff
@@ -40,6 +47,9 @@ func newTreeStoreDriver(dir, comp string) (*treeStorageDriverInstance, error) {
 	return inst, nil
 }
 
+// nopWriter is a ReadCloser that ignores the Close() call
```
Comment (bowlofeggs) [inline server/storage/tree.go]:
```suggestion
// nopWriter is a ReadCloser that ignores the Close() call.
```

File: server/storage/tree.go [hunk 14]
```diff
@@ -40,6 +47,9 @@ func newTreeStoreDriver(dir, comp string) (*treeStorageDriverInstance, error) {
...
+// it is used to make sure an innerReadCloser can be used if uncompressed storage is used.
```
Comment (bowlofeggs) [inline server/storage/tree.go]:
```suggestion
// It is used to make sure an innerReadCloser can be used if uncompressed storage is used.
```

File: server/storage/tree.go [hunk 15]
```diff
@@ -40,6 +47,9 @@ func newTreeStoreDriver(dir, comp string) (*treeStorageDriverInstance, error) {
...
+// It is basically a Writer/WriteCloser equivalent to ioutil.NopCloser
```
Comment (bowlofeggs) [inline server/storage/tree.go]:
```suggestion
// It is basically a Writer/WriteCloser equivalent to ioutil.NopCloser.
```

File: server/storage/tree.go [hunk 16]
```diff
@@ -67,6 +79,7 @@ func (d *treeStorageDriverInstance) compressReader(r io.Reader) (io.ReadCloser,
 	return ioutil.NopCloser(r), nil
 }
 
+// compressReader returns an io.ReadCloser around "r" that compresses the object per configuration
```
Comment (bowlofeggs) [inline server/storage/tree.go]:
```suggestion
// compressWriter returns an io.WriteCloser around "w" that compresses the object per configuration
```

**Review by bowlofeggs — CHANGES_REQUESTED** [review]

File: server/storage/generic.go (hunk 5 above)
Comment (puiterwijk) [reply server/storage/generic.go]: It's defined here because this set of definitions comes from https://github.com/git/git/blob/master/cache.h#L432 .
What I meant is that they should never occur returned from specific functions other than if their response value is documented as such (e.g. the `tree` driver sends `ObjectTypeBad` back from `ReadObject` when an object was not found, because in this part the `ObjectType` should not even be used, because a non-`nil` `err` is returned).
But their values should *never* occur in external (pack)files, as doing such would be a clear error.

File: server/storage/tree.go (hunk 16 above)
Comment (puiterwijk) [reply server/storage/tree.go]: Teaches me to copy/paste.

**Review by bowlofeggs — APPROVED** [review]
